tarantool/msgpack_ext/types/datetime.py

```diff
@@ -0,0 +1,237 @@
+#   little-endian order.
+# * The optional fields (8 bytes), if any of them have a non-zero value.
+#   The fields include nsec (4 bytes), tzoffset (2 bytes), and
+#   tzindex (2 bytes) packed in the little-endian order.
+#
+# seconds is seconds since Epoch, where the epoch is the point where the time
+# starts, and is platform dependent. For Unix, the epoch is January 1,
+# 1970, 00:00:00 (UTC). Tarantool uses a double type, see a structure
+# definition in src/lib/core/datetime.h and reasons in
+# https://github.com/tarantool/tarantool/wiki/Datetime-internals#intervals-in-c
+#
+# nsec is nanoseconds, fractional part of seconds. Tarantool uses int32_t, see
+# a definition in src/lib/core/datetime.h.
+#
+# tzoffset is timezone offset in minutes from UTC. Tarantool uses a int16_t type,
+# see a structure definition in src/lib/core/datetime.h.
+#
+# tzindex is Olson timezone id. Tarantool uses a int16_t type, see a structure
+# definition in src/lib/core/datetime.h. If both tzoffset and tzindex are
+# specified, tzindex has the preference and the tzoffset value is ignored.
+
+SECONDS_SIZE_BYTES  = 8
+NSEC_SIZE_BYTES     = 4
+TZOFFSET_SIZE_BYTES = 2
+TZINDEX_SIZE_BYTES  = 2
+
+BYTEORDER = 'little'
+
+NSEC_IN_SEC = 1000000000
+SEC_IN_MIN = 60
+MIN_IN_DAY = 60 * 24
+
+DATETIME_TIMEZONE_NAME_POS = 1
+
+def get_bytes_as_int(data, cursor, size):
+    part = data[cursor:cursor + size]
+    return int.from_bytes(part, BYTEORDER, signed=True), cursor + size
+
+def get_int_as_bytes(data, size):
+    return data.to_bytes(size, byteorder=BYTEORDER, signed=True)
+
+def compute_offset(timestamp):
+    if timestamp.tz is None:
+        return 0
+
+    utc_offset = timestamp.tz.utcoffset(timestamp)
+    # There is no precision loss since offset is in minutes
+    return utc_offset.days * MIN_IN_DAY + utc_offset.seconds // SEC_IN_MIN
+
+def get_python_tzname(timestamp):
+    tzinfo = timestamp.tzinfo
+
+    if tzinfo is None:
+        return None
+
+    if isinstance(tzinfo, pytz.tzinfo.BaseTzInfo):
+        return tzinfo.zone
+
+    if isinstance(tzinfo, pytz._FixedOffset):
+        # pytz.FixedOffset(0) is actually pytz.utc timezone and
+        # even not a pytz._FixedOffset.
+        return None
+
+    if isinstance(tzinfo, datetime.timezone):
+        # The only way to legally obtain datetime.timezone name is timezone.tzname(dt).
+        #
+        # But if name is not provided in the constructor, the name returned
+        # by tzname(dt) is generated from the value of the offset.
+        # We want to get `None` if name was not provided in the constructor.
+        # The only way to work with such behavior is to mess with init attributes.
+        #
+        # https://github.com/python/cpython/blob/1756ffd66a38755cd45de51316d66266ae30e132/Lib/datetime.py#L2323-L2327 
+        initargs = tzinfo.__getinitargs__()
+        if len(initargs) > DATETIME_TIMEZONE_NAME_POS:
+            return initargs[DATETIME_TIMEZONE_NAME_POS]
+        
+        return None
```

Comment:
It seems to me do not handle this case. This can cause unexpected behavior.

But up to you.

Reply:
We build custom abbreviated timezones as datetime.timezone() with name, so this case is important.